[GTK][WPE] Remove WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} in WebKitSettings.cpp

aperezdc opened the pull request:
#### 6590010e1ff81a61395344c17882942c56ceea78
<pre>
[GTK][WPE] Remove WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} in WebKitSettings.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=282308">https://bugs.webkit.org/show_bug.cgi?id=282308</a>

Reviewed by Michael Catanzaro.

Introduce helpers for GLib functions which return heap-allocated
arrays, which wrap those into a GMallocSpan (same as MallocSpan,
with GMalloc as allocator by default) that takes care of freeing
memory automatically. Then use the new utility functions to remove
WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} in WebKitSettings.cpp.

* Source/WTF/wtf/PlatformGTK.cmake: List GSpanExtras.cpp for building.
* Source/WTF/wtf/PlatformWPE.cmake: Ditto.
* Source/WTF/wtf/glib/GSpanExtras.cpp: Added. Includes new functions
WTF::gKeyFileGetKeys() and WTF::gObjectClassGetProperties(), which wrap
the corresponding GLib functions and return a GMallocSpan, easing their
usage at call sites.
(WTF::gKeyFileGetKeys): Added.
(WTF::gObjectClassGetProperties): Added.
* Source/WTF/wtf/glib/GSpanExtras.h: Define GMalloc which implements
memory operations calling GLib functions while allowing per-type
overides, to be used with MallocSpan. A convenience template alias,
GMallocSpan is also defined, which takes care of picking the
corresponding GMalloc&lt;T&gt; specialization.
(WTF::GMalloc::malloc): Added.
(WTF::GMalloc::tryMalloc): Added.
(WTF::GMalloc::zeroedMalloc): Added.
(WTF::GMalloc::tryZeroedMalloc): Added.
(WTF::GMalloc::realloc): Added.
(WTF::GMalloc::tryRealloc): Added.
(WTF::GMalloc::free): Added.
(WTF::GMalloc::nextCapacity): Added.
(WTF::GMallocStrv::free): Added.
(WTF::adoptGMallocSpan): Convenience template function that forwards to
adoptMallocSpan() picking GMalloc as the allocator by default.
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkit_settings_apply_from_key_file): Use the new helper functions.

Canonical link: <a href="https://commits.webkit.org/285949@main">https://commits.webkit.org/285949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/01e17decaf6bdb7e3c107ff3776bcf5c2e86c68b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74291 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53720 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78670 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25529 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62853 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1505 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58400 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16729 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77358 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48562 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63906 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38812 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45529 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21401 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23862 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67428 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66951 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21748 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80187 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73549 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1608 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/919 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66693 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63924 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65973 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9911 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8065 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95330 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11467 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1572 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4360 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20934 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1601 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1589 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1608 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/glib/GSpanExtras.cpp:
```diff
@@ -0,0 +1,44 @@
+
+    size_t keyCount = 0;
+    GStrv keys = g_key_file_get_keys(keyFile, groupName, &keyCount, &error.outPtr());
+    return adoptGMallocSpan<char*, GStrv>(keys, keyCount);
```
The GStrv typedef is intended for use with g_autofree. It doesn't add any value in C++; it just obscures the actual type `char**`. I've never much liked typedefs that hide pointers.

So yeah, my style suggestion is:

```suggestion
GMallocSpan<char*, char**> gKeyFileGetKeys(GKeyFile* keyFile, const char* groupName, GUniqueOutPtr<GError>& error)
{
    ASSERT(keyFile);
    ASSERT(groupName);

    size_t keyCount = 0;
    char** keys = g_key_file_get_keys(keyFile, groupName, &keyCount, &error.outPtr());
    return adoptGMallocSpan<char*, char**>(keys, keyCount);
```

mcatanzaro commented on Source/WTF/wtf/glib/GSpanExtras.h:
```diff
@@ -19,11 +19,40 @@
+
+extern "C" {
+typedef char** GStrv;
+void g_strfreev(GStrv);
```
```suggestion
void g_strfreev(char**);
```

mcatanzaro commented on Source/WTF/wtf/glib/GSpanExtras.h:
```diff
@@ -19,11 +19,40 @@
+
+template <>
+struct GMalloc<GStrv> {
+    static void free(GStrv ptr) { g_strfreev(ptr); }
```
```suggestion
struct GMalloc<char**> {
    static void free(char** ptr) { g_strfreev(ptr); }
```

aperezdc replied:
Here I would like to actually have a way of telling apart whether `g_free()` or `g_strfreev()` should be used. By specializing on `GMalloc<char**>` (or `GMalloc<GStrv>`) we would always pick the specialization that uses `g_strfreev()` when instantiating `GMallocSpan<char*>` without having the chance of picking the `g_free()` one&mdash;which may be needed if we want to wrap something returned by GLib which is a `char**` with `(transfer container)`!

So what I am going to do is plainly:

```cpp
struct GMalloc {
    static void free(void* ptr) { g_free(ptr); }
};

struct GMallocStrv {
    static void free(void* ptr) { g_strfreev(ptr); }
};

template <typename T, typename Malloc = GMalloc>
using GMallocSpan = MallocSpan<T, Malloc>;
```

This way by default a `GMallocSpan<char*>` will use `GMalloc`, and we can specifically opt-in to use `g_strfreev()` using a `GMallocSpan<char*, GMallocStrv>`.
```

mcatanzaro reviewed (approved):
My comments are just style nits. This idea seems fine.

mcatanzaro commented on Source/WTF/wtf/glib/GSpanExtras.h:
```diff
@@ -19,11 +19,40 @@
+template <typename T>
+struct GMalloc {
+    static void free(T* ptr) { g_free(ptr); }
+};
```
Maybe we should just go ahead and add malloc, zeroedMalloc, tryMalloc, tryZeroedMalloc, and realloc now, even though they're not used yet. They should all be trivial: use g_malloc(), g_malloc0(), g_try_malloc(), g_try_malloc0(), and g_realloc().

aperezdc replied:
Good idea, but I am only going to define them for `GMalloc`, leaving them out of `GMallocStrv` because for now the intention is to only use the latter for the return value of `gKeyFileGetKeys()`.

aperezdc committed:
[GTK][WPE] Remove WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} in WebKitSettings.cpp
https://bugs.webkit.org/show_bug.cgi?id=282308

Reviewed by Michael Catanzaro.

Introduce helpers for GLib functions which return heap-allocated
arrays, which wrap those into a GMallocSpan (same as MallocSpan,
with GMalloc as allocator by default) that takes care of freeing
memory automatically. Then use the new utility functions to remove
WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} in WebKitSettings.cpp.

* Source/WTF/wtf/PlatformGTK.cmake: List GSpanExtras.cpp for building.
* Source/WTF/wtf/PlatformWPE.cmake: Ditto.
* Source/WTF/wtf/glib/GSpanExtras.cpp: Added. Includes new functions
WTF::gKeyFileGetKeys() and WTF::gObjectClassGetProperties(), which wrap
the corresponding GLib functions and return a GMallocSpan, easing their
usage at call sites.
(WTF::gKeyFileGetKeys): Added.
(WTF::gObjectClassGetProperties): Added.
* Source/WTF/wtf/glib/GSpanExtras.h: Define GMalloc which implements
memory operations calling GLib functions while allowing per-type
overides, to be used with MallocSpan. A convenience template alias,
GMallocSpan is also defined, which takes care of picking the
corresponding GMalloc<T> specialization.
(WTF::GMalloc::malloc): Added.
(WTF::GMalloc::tryMalloc): Added.
(WTF::GMalloc::zeroedMalloc): Added.
(WTF::GMalloc::tryZeroedMalloc): Added.
(WTF::GMalloc::realloc): Added.
(WTF::GMalloc::tryRealloc): Added.
(WTF::GMalloc::free): Added.
(WTF::GMalloc::nextCapacity): Added.
(WTF::GMallocStrv::free): Added.
(WTF::adoptGMallocSpan): Convenience template function that forwards to
adoptMallocSpan() picking GMalloc as the allocator by default.
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkit_settings_apply_from_key_file): Use the new helper functions.

Canonical link: https://commits.webkit.org/285949@main
